### PR TITLE
Store and verify Acme account location 

### LINKED ISSUE #623: Changing ACME server host or port causes existing accounts to craft malformed requests
### Subject of the issue

**Please note, I am not sure if this is a bug.** It is just unexpected behavior I found after wasting a little of Max Furman's time (sorry about that) troubleshooting #193.

Basically, the `kid` used with JWS contain the scheme, host, port, and path (i.e. `https://localhost:9000/acme/myca/account/...`) from the ACME server's configuration at the time.

However, if the ACME server changes any part of the address, then the kid prefix check fails here: https://github.com/smallstep/certificates/blob/6476eb45a7e97ac43a37da7ab4893555625de249/acme/api/middleware.go#L324-L331

### Steps to reproduce
1. Create an ACME account with the server at one URL
2. Move the ACME server to a new URL (any of scheme, host, port, or path).
3. Try another ACME transaction: `urn:ietf:params:acme:error:malformed - The request message was malformed`

### Expected behaviour
I am not sure. Maybe that error is expected.

### Actual behaviour
`urn:ietf:params:acme:error:malformed - The request message was malformed`

### Additional context
I do think this is unexpected, especially with internal/development ACME servers, since in my case I originally had the account created at `localhost:9000` then set up a reverse proxy for the ACME server at `localhost:9001` and started pointing the client to use that one, but reuse the same account. Since the client stores the account `kid` from when it created the account, it retained the `localhost:9000` value.

Essentially, this means the ACME server can never change its address after it is being used, unless new accounts are created _or_ clients change their stored account locations to reflect the new server address. ~~Maybe that is the best way to handle it.~~ Edit: RFC 8555 suggests otherwise, see discussion below.

I am not familar enough with JWK/JWS to know the value of the prefix check, but I assume it wasn't invented for nothing. So this might very well be on the client to fix.

(Incidentally, setting STEPDEBUG=1 was not enough to emit detailed error messages. I had to add my own debug logging to find this.)

Again, I am not sure if this is a server bug. Wanted to open a discussion anyway.

### PR DESCRIPTION
**Pull request by dopey**
Fixes #623

### CODE REVIEW

File: acme/account_test.go [hunk 1]
```diff
@@ -66,6 +66,23 @@ func TestKeyToID(t *testing.T) {
 	}
 }
 
+func TestAccount_GetLocation(t *testing.T) {
+	locationPrefix := "https://test.ca.smallstep.com/acme/foo/account/"
```
Comment (maraino) [inline acme/account_test.go:70]: To properly test that the URL is the right one you should get an account for NewAccount, or at least use something like:
```go
locationPrefix := getAccountLocationPath(ctx, linker, "")
```

File: acme/account_test.go [hunk 2]
```diff
@@ -66,6 +66,23 @@ func TestKeyToID(t *testing.T) {
...
+	type test struct {
+		acc *Account
+		exp string
+	}
+	tests := map[string]test{
+		"empty":     {acc: &Account{LocationPrefix: ""}, exp: ""},
+		"not-empty": {acc: &Account{ID: "bar", LocationPrefix: locationPrefix}, exp: locationPrefix + "bar"},
```
Comment (maraino) [inline acme/account_test.go:77]: With the above comment the expected should be `https://test.ca.smallstep.com/acme/foo/account/bar`

File: acme/api/middleware_test.go [hunk 3]
```diff
@@ -754,7 +730,49 @@ func TestHandler_lookupJWK(t *testing.T) {
...
+				linker: acme.NewLinker("test.ca.smallstep.com", "acme"),
+				db: &acme.MockDB{
+					MockGetAccount: func(ctx context.Context, id string) (*acme.Account, error) {
+						assert.Equals(t, id, accID)
+						return acc, nil
+					},
+				},
+				ctx:        ctx,
+				statusCode: 401,
+				err:        acme.NewError(acme.ErrorUnauthorizedType, "kid does not matc hstored account location; expected foobar, but %q", prefix+accID),
```
Comment (maraino) [inline acme/api/middleware_test.go:747]:
```suggestion
				err:        acme.NewError(acme.ErrorUnauthorizedType, "kid does not match stored account location; expected foobar, but %q", prefix+accID),
```

File: acme/db/nosql/account_test.go
Comment (maraino) [inline acme/db/nosql/account_test.go:1]: Can we add LocationPrefix to one of the ok tests in UpdateAccount

**Review by maraino — CHANGES_REQUESTED** [review]
Looks good, but I'd like a few changes in the tests.

**Review by maraino — APPROVED** [review]
lgtm

File: acme/account.go [hunk 4]
```diff
@@ -20,6 +20,22 @@ type Account struct {
...
+func (a *Account) GetLocation() string {
+	if a.LocationPrefix == "" {
+		return ""
+	}
+	return a.LocationPrefix + a.ID
+}
+
+// GetProvisionerName returns the provisioner name extracted from the stored
+// location.
+func (a *Account) GetProvisionerName() string {
```
Comment (hslatman) [inline acme/account.go:37]: The comment seems to reflect the old logic

File: acme/api/middleware.go [hunk 5]
```diff
@@ -16,7 +17,6 @@ import (
 	"github.com/smallstep/certificates/api/render"
 	"github.com/smallstep/certificates/authority/provisioner"
 	"github.com/smallstep/certificates/logging"
-	"github.com/smallstep/nosql"
```
Comment (hslatman) [inline acme/api/middleware.go:19]: Nice 😄

File: acme/api/middleware.go [hunk 6]
```diff
@@ -324,6 +320,45 @@ func lookupJWK(next nextHTTP) nextHTTP {
...
+				// Verify that the provisioner with which the account was created
+				// matches the provisioner in the request URL.
+				reqProv := acme.MustProvisionerFromContext(ctx)
+				reqProvName := reqProv.GetName()
+				accProvName := acc.GetProvisionerName()
+				if reqProvName != accProvName {
+					// Provisioner in the URL must match the provisioner with
+					// which the account was created.
+					render.Error(w, acme.NewError(acme.ErrorUnauthorizedType,
+						"account provisioner does not match requested provisioner; account provisioner = %s, reqested provisioner = %s",
```
Comment (hslatman) [inline acme/api/middleware.go:343]:
```suggestion
						"account provisioner does not match requested provisioner; account provisioner = %s, requested provisioner = %s",
```

File: acme/db/nosql/account_test.go [hunk 7]
```diff
@@ -686,12 +708,7 @@ func TestDB_UpdateAccount(t *testing.T) {
 					assert.HasPrefix(t, err.Error(), tc.err.Error())
 				}
 			} else {
-				if assert.Nil(t, tc.err) {
-					assert.Equals(t, tc.acc.ID, dbacc.ID)
-					assert.Equals(t, tc.acc.Status, dbacc.Status)
-					assert.Equals(t, tc.acc.Contact, dbacc.Contact)
-					assert.Equals(t, tc.acc.Key.KeyID, dbacc.Key.KeyID)
-				}
```
Comment (hslatman) [inline acme/db/nosql/account_test.go:694]: Is there a reason for removing these assertions? I guess they were kind of superfluous with the assertions being done in the mocked functions?

Comment (dopey) [reply acme/db/nosql/account_test.go:694]: Exactly this. They were already being asserted from within the mock.

**Review by maraino — APPROVED** [review]
lgtm

There's no need for `GetProvisionerName()` because the ProvisionerName is public, and it doesn't have any logic.

**Review by maraino — APPROVED** [review]

**Pull request merged**